Lib/socket.py

```diff
@@ -463,6 +465,36 @@ def fromfd(fd, family, type, proto=0):
     nfd = dup(fd)
     return socket(family, type, proto, nfd)
 
+_GLOBAL_DEFAULT_MAXFDS = object()
+
+if hasattr(_socket.socket, "sendmsg"):
+    def send_fds(sock, buffers, fds, flags=0, address=None):
+        """ send_fds(sock, buffers, fds[, flags[, address]]) -> socket object
+
+        Send the list of file descriptors fds over an AF_UNIX socket.
+        """
+        return sock.sendmsg(buffers, [(_socket.SOL_SOCKET,
+            _socket.SCM_RIGHTS, array.array("i", fds))])
+    __all__.append("send_fds")
+
+if hasattr(_socket.socket, "recvmsg"):
+    def recv_fds(sock, bufsize, maxfds, flags=0):
+        """ recv_fds(sock, bufsize, maxfds[, flags]) -> (socket object, socket object)
+
+        receive up to maxfds file descriptors returning the message
+        data and a list containing the descriptors.
+        """
+        # Array of ints
+        fds = array.array("i")
+        msg, ancdata, flags, addr = sock.recvmsg(bufsize,
+            _socket.CMSG_LEN(maxfds * fds.itemsize))
```

Comment:
This will likely fail on FreeBSD as  RFC 3542 requires portable applications to use  CMSG_SPACE if I am not mistaken.

Reply:
We should check that this works on FreeBSD to avoid broken buildbots with obscure errors.

Reply:
@pablogsal You mean at test level ?

Reply:
I mean that FreeBSD systems running that code can fail and that will likely appear when running the tests that go over this code path. 

This is an example of similar scenarios of CMSG_LEN vs CMSG_SPACE:

https://github.com/python/cpython/pull/9613

Reply:
I dont have a freeBSD system to confirm this atm. When I do find, I will confirm but someone else can help confirm this. Do you have a FreeBSD to help check this?

Reply:
I will try to build this PR in the buildbots (that include FreeBSD systems).

Reply:
Here are the custom builds:

https://buildbot.python.org/all/#/builders/175/builds/39
https://buildbot.python.org/all/#/builders/171/builds/50

Reply:
It seems that my suspicions were correct:

```exception
test_errors (test.test_socket.SendfileUsingSendfileTest) ... ok
test test_socket failed
test_new_tcp_flags (test.test_socket.TestMSWindowsTCPFlags) ... skipped 'requires Windows'
======================================================================
ERROR: testSendAndRecvFds (test.test_socket.BasicTCPTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/home/buildbot/python/custom.koobs-freebsd10.nondebug/build/Lib/test/test_socket.py", line 2294, in testSendAndRecvFds
    socket.send_fds(sock, [MSG], fds)
  File "/usr/home/buildbot/python/custom.koobs-freebsd10.nondebug/build/Lib/socket.py", line 476, in send_fds
    return sock.sendmsg(buffers, [(_socket.SOL_SOCKET,
OSError: [Errno 22] Invalid argument
======================================================================
ERROR: testSendAndRecvFds (test.test_socket.BasicTCPTest2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/home/buildbot/python/custom.koobs-freebsd10.nondebug/build/Lib/test/test_socket.py", line 2294, in testSendAndRecvFds
    socket.send_fds(sock, [MSG], fds)
  File "/usr/home/buildbot/python/custom.koobs-freebsd10.nondebug/build/Lib/socket.py", line 476, in send_fds
    return sock.sendmsg(buffers, [(_socket.SOL_SOCKET,
OSError: [Errno 22] Invalid argument
----------------------------------------------------------------------
Ran 581 tests in 29.349s
```

Reply:
@koobs Could you please take a look? The new feature is failing on FreeBSD.

Reply:
I can confirm that is the same error as in OSX so it may be easier to investigate there.

Reply:
According to the FreeBSD/OSX man page:

>    Open file descriptors are now passed as ancillary data for AF_UNIX domain
     sockets, with cmsg_level set to SOL_SOCKET and cmsg_type set to
     SCM_RIGHTS.

I am not sure if they mean that is done automatically or that you should do that.